src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -25,25 +26,23 @@ private BoundExpression BindDeconstruction(AssignmentExpressionSyntax node, Diag
             var left = node.Left;
             var right = node.Right;
 
-            if (node.IsDeconstructionDeclaration())
-            {
-                return BindDeconstructionDeclaration(node, left, right, diagnostics);
-            }
-
-            if (DeconstructionIsAssignment(left))
-            {
-                var tuple = (TupleExpressionSyntax)left;
-                ArrayBuilder<DeconstructionVariable> checkedVariables = BindDeconstructionAssignmentVariables(tuple.Arguments, tuple, diagnostics);
-                var result = BindDeconstructionAssignment(node, node.Right, checkedVariables, diagnostics, isDeclaration: false);
-                FreeDeconstructionVariables(checkedVariables);
-                return result;
-            }
+            Debug.Assert(node.IsDeconstructionDeclaration() || DeconstructionIsAssignment(left));
+            return BindDeconstruction(node, left, right, diagnostics);
+        }
 
-            // TODO REVIEW underscore is currently blocked in d-declaration
-            Error(diagnostics, ErrorCode.ERR_CannotDeconstructDynamic, left);
-            return BadExpression(left);
+        internal BoundDeconstructionAssignmentOperator BindDeconstruction(CSharpSyntaxNode node, ExpressionSyntax declaration, ExpressionSyntax right,
+                                                        DiagnosticBag diagnostics, BoundDeconstructValuePlaceholder rightPlaceholder = null)
```

Comment:
`ForEachLoopBinder` makes use of that functionality, as the item to deconstruct changes with each iteration.